File: src/ApplicationTest/Rxmxnx.JNetInterface.VbApplicationTest/JCompiler.vb
Imports System.IO

<CodeAnalysis.ExcludeFromCodeCoverage>
public NotInheritable Partial Class JCompiler
    Public Property JdkPath As String
    Public Property CompilerPath As String
    Public Property LibraryPath As String

    Public Async Function CompileHelloJniClassAsync() As Task(Of Byte())
        Dim javaFilePath As String = Path.Combine(Path.GetTempPath(), "HelloDotnet.java")
        Dim classFilePath As String = Path.Combine(Path.GetTempPath(), "HelloDotnet.class")
        Try
            File.Delete(javaFilePath)
            File.Delete(classFilePath)
            Await File.WriteAllTextAsync(javaFilePath, JavaCode)
            Dim _
                info As _
                    New ProcessStartInfo(Path.Combine(Me.JdkPath, Me.CompilerPath), Chr(32) & javaFilePath & Chr(32))
            Dim javac As Process = Process.Start(info)
            Await javac.WaitForExitAsync()
            Return Await File.ReadAllBytesAsync(classFilePath)
        Finally
            File.Delete(javaFilePath)
            File.Delete(classFilePath)
        End Try
    End Function

    Public Function GetLibrary() As JVirtualMachineLibrary
        Dim result as JVirtualMachineLibrary = JVirtualMachineLibrary.LoadLibrary(Path.Combine(Me.JdkPath,
                                                                                               Me.LibraryPath))
        If result is Nothing Then
            Throw New InvalidOperationException("Invalid JVM library.")
        End If
        Return result
    End Function
End Class

File: src/ApplicationTest/Rxmxnx.JNetInterface.ConsoleTest/JCompiler.cs
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;

namespace Rxmxnx.JNetInterface.ApplicationTest;

/// <summary>
/// Java compiler.
/// </summary>
[ExcludeFromCodeCoverage]
public sealed partial class JCompiler
{
	/// <summary>
	/// Sdk path.
	/// </summary>
	public String JdkPath { get; set; } = default!;
	/// <summary>
	/// JAVAC path.
	/// </summary>
	public String CompilerPath { get; set; } = default!;
	/// <summary>
	/// JVM library path.
	/// </summary>
	public String LibraryPath { get; set; } = default!;

	/// <summary>
	/// Compiles HelloJni java class.
	/// </summary>
	/// <returns>Java class bytecode.</returns>
	public async Task<Byte[]> CompileHelloJniClassAsync()
	{
		String javaFilePath = Path.Combine(Path.GetTempPath(), "HelloDotnet.java");
		String classFilePath = Path.Combine(Path.GetTempPath(), "HelloDotnet.class");
		try
		{
			File.Delete(javaFilePath);
			File.Delete(classFilePath);
			await File.WriteAllTextAsync(javaFilePath, JCompiler.JavaCode);
			ProcessStartInfo info = new(Path.Combine(this.JdkPath, this.CompilerPath))
			{
				ArgumentList = { javaFilePath, },
			};
			Process javac = Process.Start(info)!;
			await javac.WaitForExitAsync();
			return await File.ReadAllBytesAsync(classFilePath);
		}
		finally
		{
			File.Delete(javaFilePath);
			File.Delete(classFilePath);
		}
	}
	/// <summary>
	/// Retrieves a <see cref="JVirtualMachineLibrary"/> instance
	/// from current JDK.
	/// </summary>
	/// <returns>A <see cref="JVirtualMachineLibrary"/> instance.</returns>
	public JVirtualMachineLibrary GetLibrary()
	{
		String libraryPath = Path.Combine(this.JdkPath, this.LibraryPath);
		return JVirtualMachineLibrary.LoadLibrary(libraryPath) ??
			throw new InvalidOperationException($"Invalid JVM library. {libraryPath}");
	}
}